### PR TITLE
map: do not allocate on lookup when key doesn't exist

### LINKED ISSUE #1516: map Lookup allocates when key does not exist
### Describe the bug

In a hot loop of Lookup where a lot of keys do not exist, the `fmt.Errorf` at https://github.com/cilium/ebpf/blob/fbb9ed862764a07745604fc608dc09534e76cd36/map.go#L698 can allocate a decent chunk of memory

### How to reproduce

1. Create map
2. Do a bunch of lookups for keys that do not exist
3. Profile allocations

### Version information

github.com/cilium/ebpf v0.15.1-0.20240709101333-5976561b28aa

### PR DESCRIPTION
**Pull request by brycekahle**
Fixes #1516

### CODE REVIEW

**Review by florianl — APPROVED** [review]

File: map_test.go [hunk 1]
```diff
@@ -245,6 +245,13 @@ func TestMapLookupKeyTooSmall(t *testing.T) {
 	qt.Assert(t, qt.IsNotNil(m.Lookup(uint32(0), &small)))
 }
 
+func TestMapLookupKeyNotFoundAllocations(t *testing.T) {
+	allocs := testing.AllocsPerRun(5, func() {
+		_ = wrapMapLookupError(unix.ENOENT)
```
Comment (lmb) [inline map_test.go:250]: Can we do this on Map.Lookup instead?

File: syscalls.go [hunk 2]
```diff
@@ -145,6 +145,13 @@ var haveNoPreallocMaps = internal.NewFeatureTest("prealloc maps", "4.6", func()
 	return nil
 })
 
+func wrapMapLookupError(err error) error {
```
Comment (lmb) [inline syscalls.go:148]: Nit: it'd be ok to inline this into `Map.lookup` I think.

File: map_test.go (hunk 1 above)
Comment (brycekahle) [reply map_test.go]: Done.

File: syscalls.go (hunk 2 above)
Comment (brycekahle) [reply syscalls.go]: Done.

**Review by lmb — APPROVED** [review]

**Pull request merged**